File: src/main/java/com/atsuishio/superbwarfare/network/message/receive/SoundClientMessage.java
package com.atsuishio.superbwarfare.network.message.receive;

import com.atsuishio.superbwarfare.Mod;
import com.atsuishio.superbwarfare.network.ClientPacketHandler;
import io.netty.buffer.ByteBuf;
import net.minecraft.core.UUIDUtil;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import org.jetbrains.annotations.NotNull;
import org.joml.Vector3f;

import java.util.UUID;

public record SoundClientMessage(
        ResourceLocation location,
        Vector3f pos,
        float radius,
        float pitch,
        UUID uuid
) implements CustomPacketPayload {

    public static final Type<SoundClientMessage> TYPE = new Type<>(Mod.loc("sound_client"));

    public static final StreamCodec<ByteBuf, SoundClientMessage> STREAM_CODEC = StreamCodec.composite(
            ResourceLocation.STREAM_CODEC,
            SoundClientMessage::location,
            ByteBufCodecs.VECTOR3F,
            SoundClientMessage::pos,
            ByteBufCodecs.FLOAT,
            SoundClientMessage::radius,
            ByteBufCodecs.FLOAT,
            SoundClientMessage::pitch,
            UUIDUtil.STREAM_CODEC,
            SoundClientMessage::uuid,
            SoundClientMessage::new
    );

    public static void handler(SoundClientMessage message) {
        ClientPacketHandler.handleSoundClientMessage(message);
    }

    @Override
    public @NotNull Type<? extends CustomPacketPayload> type() {
        return TYPE;
    }
}

File: src/main/java/com/atsuishio/superbwarfare/network/message/receive/SoundClientMessage.kt
package com.atsuishio.superbwarfare.network.message.receive

import com.atsuishio.superbwarfare.event.ClientEventHandler
import com.atsuishio.superbwarfare.network.ClientPacketPayload
import com.atsuishio.superbwarfare.network.PayloadContext
import com.atsuishio.superbwarfare.tools.localPlayer
import com.atsuishio.superbwarfare.tools.options
import com.atsuishio.superbwarfare.tools.queueClientWorkIfDelayed
import kotlinx.serialization.Contextual
import kotlinx.serialization.Serializable
import net.minecraft.client.CameraType
import net.minecraft.resources.ResourceLocation
import net.minecraft.sounds.SoundEvent
import net.minecraft.sounds.SoundSource
import net.minecraft.world.phys.Vec3
import java.util.*

@Serializable
data class SoundClientMessage(
    @Contextual val location: ResourceLocation,
    val x: Double,
    val y: Double,
    val z: Double,
    val radius: Float,
    val pitch: Float,
    @Contextual val sender: UUID,
) : ClientPacketPayload() {

    override fun PayloadContext.handler() {
        val player = localPlayer ?: return
        if (player.getUUID() == sender && (options.cameraType == CameraType.FIRST_PERSON || ClientEventHandler.zoomVehicle)) return

        val sound = SoundEvent.createVariableRangeEvent(location)
        val distance = player.position().distanceTo(Vec3(x, y, z))

        queueClientWorkIfDelayed((distance / 17).toInt()) {
            player.level().playSound(player, x, y, z, sound, SoundSource.BLOCKS, radius, pitch)
        }
    }
}
